transform/event_origin_filter.go

```diff
@@ -0,0 +1,99 @@
+package transform
+
+import (
+	"fmt"
+
+	"github.com/streamingfast/bstream"
+	"github.com/streamingfast/bstream/transform"
+	"github.com/streamingfast/dstore"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	pbtransform "github.com/figment-networks/proto-cosmos/pb/sf/cosmos/transform/v1"
+	pbcosmos "github.com/figment-networks/proto-cosmos/pb/sf/cosmos/type/v1"
+)
+
+var EventOriginFilterMessageName = proto.MessageName(&pbtransform.EventOriginFilter{})
+
+func EventOriginFilterFactory(indexStore dstore.Store, possibleIndexSizes []uint64) *transform.Factory {
+	return &transform.Factory{
+		Obj: &pbtransform.EventOriginFilter{},
+		NewFunc: func(message *anypb.Any) (transform.Transform, error) {
+			if message.MessageName() != EventOriginFilterMessageName {
+				return nil, fmt.Errorf("expected type url %q, received %q", EventOriginFilterMessageName, message.TypeUrl)
+			}
+
+			filter := &pbtransform.EventOriginFilter{}
+			err := proto.Unmarshal(message.Value, filter)
+			if err != nil {
+				return nil, fmt.Errorf("unexpected unmarshal error: %w", err)
+			}
+
+			if len(filter.EventOrigins) == 0 {
+				return nil, fmt.Errorf("event origin filter requires at least one event origin")
+			}
+
+			eventOriginMap := make(map[string]bool)
+			for _, acc := range filter.EventOrigins {
+				eventOriginMap[acc] = true
+			}
+
+			return &EventOriginFilter{
+				EventOrigins:       eventOriginMap,
+				possibleIndexSizes: possibleIndexSizes,
+				indexStore:         indexStore,
+			}, nil
+		},
+	}
+}
+
+type EventOriginFilter struct {
+	EventOrigins map[string]bool
+
+	indexStore         dstore.Store
+	possibleIndexSizes []uint64
+}
+
+func (p *EventOriginFilter) String() string {
+	return fmt.Sprintf("%v", p.EventOrigins)
+}
+
+func (p *EventOriginFilter) Transform(readOnlyBlk *bstream.Block, in transform.Input) (transform.Output, error) {
+	block := p.filterEventsOrigins(readOnlyBlk.ToProtocol().(*pbcosmos.Block))
+
+	return block, nil
+}
+
+func (p *EventOriginFilter) filterEventsOrigins(block *pbcosmos.Block) *pbcosmos.Block {
+
+	// if filter doesn't pass these Event Origins, nullify the objects in the block
+	if !p.EventOrigins["BeginBlock"] {
+		block.ResultBeginBlock.Events = []*pbcosmos.Event{}
+	}
+	if !p.EventOrigins["EndBlock"] {
+		block.ResultEndBlock.Events = []*pbcosmos.Event{}
+	}
+	if !p.EventOrigins["DeliverTx"] {
+		for _, tx := range block.Transactions {
+			tx.Result.Events = []*pbcosmos.Event{}
+		}
+	}
```

Comment:
Let's put this directly in the `Transform` method and remove `filterEventsOrigins`.